sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -79,12 +81,669 @@ object GpuHashJoin {
+      if (mask != null) {
+        mask.close()
+        mask = null
+      }
+    }
+  }
+}
+
+/**
+ * Generic trait for all join gather instances.
+ * All instances should be spillable.
+ * The life cycle of this assumes that when it is created that the data and
+ * gather maps will be used shortly.
+ * If you are not going to use these for a while, like when returning from an iterator,
+ * then allowSpilling should be called so that the cached data is released and spilling
+ * can be allowed.  If you need/want to use the data again, just start using it, and it
+ * will be cached yet again until allowSpilling is called.
+ * When you are completely done with this object call close on it.
+ */
+trait JoinGatherer extends AutoCloseable with Arm {
+  /**
+   * Gather the next n rows from the join gather maps.
+   * @param n how many rows to gather
+   * @return the gathered data as a ColumnarBatch
+   */
+  def gatherNext(n: Int): ColumnarBatch
+
+  /**
+   * Is all of the data gathered so far.
+   */
+  def isDone: Boolean
+
+  /**
+   * Number of rows left to gather
+   */
+  def numRowsLeft: Long
+
+  /**
+   * Indicate that we are done messing with the data for now and it can be spilled.
+   */
+  def allowSpilling(): Unit
+
+  /**
+   * A really fast and dirty way to estimate the size of each row in the join output
+   */
+  def realCheapPerRowSizeEstimate: Double
+
+  /**
+   * Get the bit count size map for the next n rows to be gathered.
+   */
+  def getBitSizeMap(n: Int): ColumnView
+
+  /**
+   * If the data is all fixed width return the size of each row, otherwise return null.
+   */
+  def getFixedWidthBitSize: Option[Int]
+
+  /**
+   * Do a complete/expensive job to get the number of rows that can be gathered to get close
+   * to the targetSize for the final output.
+   */
+  def gatherRowEstimate(targetSize: Long): Int = {
+    val bitSizePerRow = getFixedWidthBitSize
+    if (bitSizePerRow.isDefined) {
+      Math.min(Math.min((targetSize/bitSizePerRow.get) / 8, numRowsLeft), Integer.MAX_VALUE).toInt
+    } else {
+      val estimatedRows = Math.min(
+        ((targetSize / realCheapPerRowSizeEstimate) * 1.1).toLong,
+        numRowsLeft)
+      val numRowsToProbe = Math.min(estimatedRows, Integer.MAX_VALUE).toInt
+      if (numRowsToProbe <= 0) {
+        1
+      } else {
+        val sum = withResource(getBitSizeMap(numRowsToProbe)) { bitSizes =>
+          bitSizes.prefixSum()
+        }
+        val cutoff = withResource(sum) { sum =>
+          withResource(new Table(sum)) { sumTable =>
+            withResource(ai.rapids.cudf.ColumnVector.fromLongs(targetSize * 8)) { bound =>
+              withResource(new Table(bound)) { boundTab =>
+                sumTable.lowerBound(boundTab, OrderByArg.asc(0))
+              }
+            }
+          }
+        }
+        withResource(cutoff) { cutoff =>
+          withResource(cutoff.copyToHost()) { hostCutoff =>
+            hostCutoff.getInt(0)
+            hostCutoff.getInt(0)
```

Comment:
It seems a little odd to have two `getInt(0)`'s here.

Reply:
ctrl-d in InteliJ duplicates a line. I hit it on accident some times and do not notice, just wonder why my copy didn't work.